### PR TITLE
bpo-35959: Rework integer overflow path in math.prod and add more tests

### PR DESCRIPTION
**Pull request by pablogsal**
The overflow check was relying on undefined behaviour as it was using the result of the multiplication to do the check, and once the overflow has already happened, any operation on the result is undefined behaviour.

I have moved the test to the correct class as well (TestMath).

https://bugs.python.org/issue35959

### CODE REVIEW

File: Lib/test/test_math.py [hunk 1]
```diff
@@ -1595,6 +1595,92 @@ def test_mtestfile(self):
...
+        self.assertEqual(prod([1.0, 2.0, 3.0, 4.0, 5.0]), 120.0)
+        self.assertEqual(prod([1, 2, 3, 4.0, 5.0]), 120.0)
+        self.assertEqual(prod([1.0, 2.0, 3.0, 4, 5]), 120.0)
+
+        # Test overflow in fast-path for integers
+        self.assertEqual(prod([1, 1, 2**32, 1, 1]), 2**32)
+        # Test overflow in fast-path for floats
+        self.assertEqual(prod([1.0, 1.0, 2**32, 1, 1]), float(2**32))
+
+        self.assertRaises(TypeError, prod)
```
Comment (eamanu) [inline Lib/test/test_math.py:1616]: Why don't write this whole self.assertRaises in an with statement.?

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -2561,8 +2561,8 @@ math_prod_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 long x = i_result * b;
                 /* Continue if there is no overflow */
                 if (overflow == 0
-                    && x < LONG_MAX && x > LONG_MIN
-                    && !(b != 0 && x / b != i_result)) {
+                    && !(b != 0 && (i_result > LONG_MAX / b
```
Comment (mdickinson) [inline Modules/mathmodule.c]: Does this code work in the case that `b` is negative?

File: Modules/mathmodule.c [hunk 3]
```diff
@@ -2561,8 +2561,8 @@ math_prod_impl(PyObject *module, PyObject *iterable, PyObject *start)
...
+                                || i_result < LONG_MIN / b))) {
```
Comment (mdickinson) [inline Modules/mathmodule.c]: There's a nasty corner case here: `LONG_MIN / b` gives undefined behaviour on a typical two's complement machine when `b = -1`.

Comment (pablogsal) [reply Modules/mathmodule.c]: Should we adopt a similar strategy as you pointed out in https://github.com/python/cpython/blob/2f1a317d5fdc45b9d714b067906f612f636ba08e/Objects/intobject.c#L496-L518 ?

Comment (mdickinson) [reply Modules/mathmodule.c]: That's one possible solution; it's not my favourite, but it has the advantage of being well battle-tested.

The Python 2.7 int * int overflow detection needs to be a little bit stricter than what you need here, though, in that here it's okay if you have the occasional false positive (overflow detected even though we didn't actually strictly overflow), while in the Python 2.7 code the overflow detection needs to be exact.

So you might be able to get away with something weaker.

File: Modules/mathmodule.c (hunk 2 above)
Comment (pablogsal) [reply Modules/mathmodule.c]: Hummm.... Unless I am missing something, it does:

```python
import math, functools, operator
x = [1,2,3,4,-1 * (1<<1000),3,4,5,6,7,]
>>> functools.reduce(operator.mul, x) == math.prod(x)
True
```

File: Modules/mathmodule.c (hunk 3 above)
Comment (pablogsal) [reply Modules/mathmodule.c]: For now, and until a better alternative is suggested I have used Python2's approach.

**Comment by pablogsal** [comment]
@mdickinson Do you have some suggestions for extra checks to cover these corner cases?

**Review by eamanu — APPROVED** [review]
LGTM. I am testing the PR on Unbuntu 18 04  x86_64 and run ok.

File: Lib/test/test_math.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_math.py:1616]: I am not sure I understand what do you mean. :(

Comment (eamanu) [reply Lib/test/test_math.py:1616]: Sorry, I mean why don't write:

```
with self.assertRaises(TypeError):
    prod()
    prod(42)
    prod(['a', 'b', 'c'])
    ...
```

File: Modules/mathmodule.c (hunk 2 above)
Comment (mdickinson) [reply Modules/mathmodule.c]: Let me rephrase: this code _doesn't_ work in the case that `b` is negative. :-)  The sense of the comparisons needs to be reversed in that case.

From another comment:

> @mdickinson Do you have some suggestions for extra checks to cover these corner cases?

I'd start out by declaring `LONG_MIN` an overflow case (even though it isn't), and assume that `LONG_MIN + LONG_MAX = -1` (which isn't guaranteed by C, but is overwhelmingly likely to be true on any hardware that Python cares about; the assumption is probably worth a comment in the source, but it's not worth worrying about beyond that). Then we just have to keep track of whether we've overflowed `LONG_MAX` in absolute value, and that's much simpler to do; it also takes care of the dividing `LONG_MIN` by `-1` problem. (This is where we win over Python 2's int * int: in that case, it's important not to treat a result of `LONG_MIN` as an overflow case, but here we don't care.)

But re-using the Python 2 overflow detection works, too.

File: Lib/test/test_math.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_math.py:1616]: Oh, I see what you mean now. These tests were not modified in this PR so I did not modify them (apart from moving them to the correct place).

Comment (eamanu) [reply Lib/test/test_math.py:1616]: ahh ok, I understand.

File: Modules/mathmodule.c (hunk 2 above)
Comment (pablogsal) [reply Modules/mathmodule.c]: @mdickinson Thanks for the explanation!

I had changed the code to use Python2's handling of the overflow because I prefer to go with battle-tested checks given how tricky multiplication overflow is. In the future, we could relax the test to use `LONG_MAX` and `LONG_MIN` if you think that is ok :)

I prefer to go with correctness and safety first, and then we can consider optimizations later.

**Comment by pablogsal** [comment]
@mdickinson At the end I have used the Python2's approach for overflow handling. Could you review again this PR?

**Comment by mdickinson** [comment]
> Could you review again this PR?

Will do, but I'm not going to have time before the weekend. If another core dev wants to review before then, that's fine with me.

**Comment by pablogsal** [comment]
I am going to merge this one as this is a known problem and I do not want to miss more alphas. As the code used is the code of how Python2 managed overflow for multiplication and this is battle tested, this should be more than ok. In case we want to improve/speedup how we handle the overflow we can do it in another PR.
